autodiff/mixed/expressions/binary.hpp

```diff
@@ -0,0 +1,191 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+#include <autodiff/mixed/expressions/expression.hpp>
+
+namespace autodiff::taperep {
+/// @brief Binary expression base class.
+/// @tparam N Maximal derivative order of expression.
+template<std::size_t N>
+struct binary_expression : expression<N>
+{
+    /// @brief Construct sum expression from left and right expression values.
+    /// @param v Value of binary expression.
+    /// @param l Value of left expression.
+    /// @param r Value of right expression.
+    explicit binary_expression(double v, const double& l, const double& r) : 
+        expression<N>(v), m_left { l }, m_right { r } { }
+
+    /// @brief Left expression value.
+    /// @return Constant reference to left expression value.
+    auto left() const -> const double&
+    {
+        return m_left;
+    }
+
+    /// @brief Right expression value.
+    /// @return Constant reference to right expression value.
+    auto right() const -> const double&
+    {
+        return m_right;
+    }
+private:
+    const double& m_left;  ///< Value reference to left part of binary node.
+    const double& m_right; ///< Value reference to right part of binary node.
+};
+
+/// @brief Binary sum expression.
+/// @tparam N Maximal derivative order of expression.
+template<std::size_t N>
+struct sum_expression final : binary_expression<N>
+{
+    using base = binary_expression<N>;
+
+    /// @brief Construct sum expression from left and right expression values.
+    /// @param l Value of left expression.
+    /// @param r Value of right expression.
+    explicit sum_expression(const double& l, const double& r) : 
+        base(l + r, l, r) 
+    { 
+        this->template partial_derivatives<first_order>() = { 1.0, 1.0 };
+
+        if constexpr (N > first_order)
+            this->template partial_derivatives<second_order>() = { };
+    }
+
+    /// @brief Update value of sum expression.
+    /// @details Values of partial derivatives are always same for sum, 
+    /// so we don't need to update them.
+    virtual void update()
+    {
+        this->value() = base::left() + base::right();
+    }
+};
+
+/// @brief Binary minus expression.
+/// @tparam N Maximal derivative order of expression.
+template<std::size_t N>
+struct minus_expression final : binary_expression<N>
```

Comment:
Here, we should use `subtraction_expression` if the intent is `x - y`. I think `minus_expression` makes sense for the unary operation `-x`.